core/commands/dag/dag.go

```diff
@@ -258,6 +273,267 @@ var DagResolveCmd = &cmds.Command{
+  If the combination of blocks from the imported CAR files and what is
+  currently present in the blockstore does not represent a complete DAG,
+  pinning of that individual root will fail.
+
+Maximum supported CAR version: 1
+`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.FileArg("path", true, true, "The path of a .car file.").EnableStdin(),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption(silentOptionName, "No output."),
+		cmds.BoolOption(pinRootsOptionName, "Pin optional roots listed in the .car headers after importing.").WithDefault(true),
+	},
+	Type: CarImportOutput{},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+
+		node, err := cmdenv.GetNode(env)
+		if err != nil {
+			return err
+		}
+
+		api, err := cmdenv.GetApi(env, req)
+		if err != nil {
+			return err
+		}
+
+		// on import ensure we do not reach out to the network for any reason
+		// if a pin based on what is imported + what is in the blockstore
+		// isn't possible: tough luck
+		api, err = api.WithOptions(options.Api.Offline(true))
+		if err != nil {
+			return err
+		}
+
+		// grab a pinlock ( which doubles as a GC lock ) so that regardless of the
+		// size of the streamed-in cars nothing will disappear on us before we had
+		// a chance to roots that may show up at the very end
+		// This is especially important for use cases like dagger:
+		//    ipfs dag import $( ... | ipfs-dagger --stdout=carfifos )
+		//
+		unlocker := node.Blockstore.PinLock()
+		defer unlocker.Unlock()
+
+		doPinRoots, _ := req.Options[pinRootsOptionName].(bool)
+
+		retCh := make(chan importResult, 1)
+		go importWorker(req, res, api, retCh)
+
+		done := <-retCh
+		if done.err != nil {
+			return done.err
+		}
+
+		// It is not guaranteed that a root in a header is actually present in the same ( or any )
+		// .car file. This is the case in version 1, and ideally in further versions too
+		// Accumulate any root CID seen in a header, and supplement its actual node if/when encountered
+		// We will attempt a pin *only* at the end in case all car files were well formed
+		//
+		// The boolean value indicates whether we have encountered the root within the car file's
+		roots := done.roots
+
+		// opportunistic pinning: try whatever sticks
+		if doPinRoots {
+
+			var failedPins int
+			for c, seen := range roots {
+
+				// We need to re-retrieve a block, convert it to ipld, and feed it
+				// to the Pinning interface, sigh...
+				//
+				// If we didn't have the problem of inability to take multiple pinlocks,
+				// we could use the Api directly like so (though internally it does the same):
+				//
+				// // not ideal, but the pinning api takes only paths :(
+				// rp := path.NewResolvedPath(
+				// 	ipfspath.FromCid(c),
+				// 	c,
+				// 	c,
+				// 	"",
+				// )
+				//
+				// if err := api.Pin().Add(req.Context, rp, options.Pin.Recursive(true)); err != nil {
+
+				ret := RootMeta{Cid: c, PresentInImport: seen}
```

Comment:
Continuing conversation from https://github.com/ipfs/go-ipfs/pull/7038#discussion_r399571054 after rebase.

Reply:
> Yes - "present in one of the supplied cars" is exactly what I was after. Would naming it PresentInImport fix the ambiguity?

No, it's still a lie. We're pinning the _DAG_, not just the roots. To me, `PresentInImport` indicates that the _DAG_ is present, not that the _root_ is present. The error message reinforces this by saying that the "data" was/wasn't present.

As a user, if I ever saw that error message and/or this type definition, I'd assume that the CAR contained all the data I tried to import. Given that we're not actually validating the CAR, this is not good.

At best, this feature is a heuristic to indicate that the CAR might have contained the DAG.

If we're going to have a feature like this, we need to implement it correctly and validate that the import contained _all_ blocks referenced by the root. However, that could force us to keep quite a few CIDs in memory so I'd definitely punt that to a future patch.

Reply:
TL;DR: I don't want to give users a false sense of security by including a sanity check that only sort of works. I realize this was helpful in debugging this command, but it only catches _one_ way in which a CAR file can be broken.

Reply:
Punted to https://github.com/ipfs/go-ipfs/issues/7120